src/Reflection/ReflectionFunctionAbstract.php

```diff
@@ -619,4 +619,17 @@ public function getReturnStatementsAst()
 
         return $visitor->getReturnNodes();
     }
+
+    private function findTypeFromAst($namespace, $locatedSource, $type)
+    {
+        $objectType = (new FindTypeFromAst())->__invoke(
+            $this->reflector->getContextFactory()->createForNamespace(
```

Comment:
We previously had it so `SourceLocator` returned a `LocatedSource`, but the suggestion was that the `LocatedSource` would definitely contain the expected source (and this is not always the case); so we mulled over things like a `PotentiallyLocatedSource` and so on, but in the end we decided that `SourceLocator` simply returning a reflection, if located, was simpler and cleaner.

An interesting thought: what if one method of loading source (e.g. a `FooSourceLocator`) needs a different way of determining context than another? May not be useful, but injecting the `ContextFactory` at the `SourceLocator` might make more sense?